dandi/support/iterators.py

```diff
@@ -0,0 +1,116 @@
+      It is a final value if `finished` is True.
+    .finished: bool
+      Set to True upon completion of iteration
+    .exc: BaseException or None
+      If not None -- the exception which was raised
+
+    Example
+    -------
+
+    Very simplistic example, since typically (not range) it would be taking some time to
+    iterate for the nested iteration::
+
+        it = IteratorWithAggregation(range(3), lambda v, t=0: v+t)
+        for v in it:
+            print(it.total, it.finished, v)
+            sleep(0.02)  # doing smth heavy, but we would know .total as soon as it is known
+
+    would produce (so 3 is known right away, again since it is just range)
+
+        3 True 0
+        3 True 1
+        3 True 2
+
+    """
+
+    def __init__(self, gen, agg, reraise_immediately=False):
+        """
+
+        Parameters
+        ----------
+        gen: iterable
+          Generator (but could be any iterable, but it would not make much sense)
+          to yield from
+        agg: callable
+          A callable with two args: new_value[, total=None] which should return adjusted
+          total. Upon first iteration, no prior `total` is provided
+        reraise_immediately: bool, optional
+          If True, it would stop yielding values as soon as it detects that some
+          exception has occurred (although there might still be values in the queue to be yielded
+          which were collected before the exception was raised)
+        """
+        self.gen = gen
+        self.agg = agg
+        self.reraise_immediately = reraise_immediately
+
+        self.total = None
+        self.finished = None
+        self._exc = None
+
+    def __iter__(self):
+        self.finished = False
+        self._exc = None
+
+        queue = Queue()
+
+        def worker():
+            """That is the one which interrogates gen and places total
+             into queue_total upon completion"""
+            total = None
+            try:
+                for value in self.gen:
+                    queue.put(value)
+                    self.total = total = (
+                        self.agg(value, total) if total is not None else self.agg(value)
+                    )
+            except BaseException as e:
+                self._exc = e
+            finally:
+                self.finished = True
+
+        t = Thread(target=worker)
+        t.start()
+
+        # yield from the queue (.total and .finished could be accessed meanwhile)
+        while True:
+            # race condition HERE between checking for self.finished and
+            if self.finished and queue.empty():
+                break
+            # in general queue should not be empty, but if it is, e.g. due to race
+            # condition with above check
+            try:
+                yield queue.get(timeout=0.001)
+            except Empty:
+                continue
+            if self.reraise_immediately and self._exc is not None:
```

Comment:
Wouldn't it be better to move this check to the top of the `while` loop in case `get` times out while an exception is raised?

Reply:
indeed! done in e52d728d9294006eedc6a9dd95e0e582f04c45bd